src/ccstruct/tabletransfer.h

```diff
@@ -0,0 +1,66 @@
+/******************************************************************************
+ * File:        tabletransfer.h
+ * Description: Infrastructure for the transfer of table detection results
+ * Author:      Stefan Brechtken
+ *
+ * (C) Copyright 2021, Stefan Brechtken
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ****************************************************************************/
+
+#ifndef TESSERACT_CCSTRUCT_TABLETRANSFER_H_
+#define TESSERACT_CCSTRUCT_TABLETRANSFER_H_
+#include <memory>
+#include <vector>
+#include "rect.h"
+
+namespace tesseract {
+
+/// Structure for data transfer from table detector 
+struct TessTable {
+  tesseract::TBOX box;
+  std::vector<tesseract::TBOX> rows;
+  std::vector<tesseract::TBOX> cols;
+};
+
+/** \brief You can use this small template function to ensure that one and
+ *   only one object of type T exists. It implements the Singleton Pattern.
+ *
+ * T must be default-constructable.
+ * Usage examples:
+ *   A& a = uniqueInstance<A>();
+ *   a.xyz();
+ *   uniqueInstance<A>(make_unique<A>(42)); // replace instance
+ *   a.foo();
+ * or
+ *   uniqueInstance<A>().xyz();
+ */
+template<typename T>
+T& uniqueInstance(std::unique_ptr<T> new_instance = nullptr)
+{
+  static std::unique_ptr<T> _instance = std::make_unique<T>();
```

Comment:
@Sintun, there is another regression caused by this pull request. It was reported [here](https://github.com/OCR-D/ocrd_all/pull/265#issuecomment-884329872). The problem can be reduced to code like this:

```
[...]
tesseract::TessBaseAPI api;
int main(int argc, char **argv) {
  api.End();
  [...]
}
```

That code creates a global `TessBaseAPI` object. Running `End()` executes `uniqueInstance<std::vector<TessTable>>().clear()` and creates a static `std::unique_ptr<std::vector<TessTable>>` object. That object is destroyed at program exit, but used again when the `TessBaseAPI` object is destroyed, too, which also calls `End()`. This results in a segmentation fault at program exit.

The bug occurs for all users of the Python wrapper `tesserocr`.

Reply:
I did not notice this whole PR until I saw the segfaults in tesserocr, tried [a workaround there](https://github.com/sirfz/tesserocr/pull/265#pullrequestreview-712376480), and had you point me here.

Should be interesting to review.

@stweil @egorpugin so what's the status right now, is 4df822a3fc481f026250fcf2bb4c0e7583df1dd5 fixing the segfaults? Does this PR need to be reopened?

Reply:
This PR was now reverted, so we need a new implementation to access the table detector results.

Reply:
And yes, the segfaults (and a 2nd regression which caused an assertion for images with tables) are fixed now.

Reply:
> This PR was now reverted, so we need a new implementation to access the table detector results.

So how to proceed? Reviews and comments here are helpful, but a new PR is needed technically...

BTW, the first action should be to add new test cases around the segfaults.